### ISSUE TITLE
Traversal algorithms do not respect node equality

### ISSUE THREAD

**Issue by cervengoc** [morelinq/MoreLINQ#249]
The two traversal algorithm implemented in MoreLinq do not respect node quality, so the result may include a node multiple times, and the algorithm is sensitive to cycles.

I'd happily put together a PR to fix this, but I was wondering if this behavior was by design, so please give your thoughts on this.

**Comment by fsateler** [comment]
Are you referring to `TraverseBreadthFirst` and `TraverseDepthFirst` ? If so, the behavior is by design (and documented). Trees can be large, and if they really are trees (ie, no loops), then doing the bookkeeping the seen nodes is an unnecessary expense.

OTOH, sometimes the source is not really a tree, so this bookkeeping is necessary. How do you propose it can be done, without burdening the users of tree-like structures?

**Comment by cervengoc** [comment]
You're right, but I think the overhead added by the tracking is not significant. If we use a `HashSet`, then adding an item and checking existence of an item both have constant complexity, which means that the overall complexity will remain linear.

But if you worry about it, writing a unit test to measure performance difference on a huge tree is not very difficult.

**Comment by fsateler** [comment]
Indeed I worry. The bad performance from bookkeeping comes from at least two dimensions:

1. Memory usage. As per [this StackOverflow post](http://stackoverflow.com/a/24942014/1117815), a HashSet uses 12 bytes per slot + the actual data. Not only this increases memory usage and pressure in the application, but it requires frequent (at least at first) allocations. Allocations can be slow. This would be particularly damaging if the node type is a struct.
2. CPU usage. It is not a given that all node classes will implement `GetHashCode` and `IEquatable` correctly. A HashSet of such items is not really fast.

For this reason I would prefer if this bookkeeping would not be done unless the user requested it.

**Comment by cervengoc** [comment]
Thank you for the comments, you have definitely a higher level of insight in this topic. So I'd suggest introducing an optional argument which controls this behavior.

**Comment by fsateler** [comment]
I don't think we should be adding this to the Traverse methods. Maybe a separate method should be added that does this. I've retitled the issue accordingly.
